File: backend/app/services/seeds.py
"""Сиды: справочники, демо-данные, импорт существующих `package_list.txt`."""

from __future__ import annotations

from datetime import timedelta

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.security import hash_password
from app.db.base import utcnow
from app.db.models import (
    License,
    ModerationRequest,
    PackageManager,
    RequestItem,
    User,
)
from app.managers.registry import all_plugins, get_plugin
from app.pipeline.runner import ensure_steps, recompute_request_status
from app.services import audit, packages
from app.services.policies import get_license_policy

DEMO_USERS = [
    ("dev.ivanov", "Иванов Иван", ["developer"], "dev.ivanov@example.com"),
    ("sec.petrov", "Петров Пётр", ["devsecops"], "sec.petrov@example.com"),
    ("legal.sidorova", "Сидорова Анна", ["legal"], "legal.sidorova@example.com"),
    ("admin", "Администратор сервиса", ["admin"], "admin@example.com"),
]


def seed_managers(session: Session) -> int:
    created = 0
    for plugin in all_plugins():
        row = session.execute(
            select(PackageManager).where(PackageManager.code == plugin.code)
        ).scalars().first()
        if row is None:
            session.add(
                PackageManager(
                    code=plugin.code,
                    title=plugin.title,
                    entry_format=plugin.entry_format,
                    enabled=True,
                )
            )
            created += 1
        else:
            row.title = plugin.title
            row.entry_format = plugin.entry_format
    session.flush()
    return created


def seed_licenses(session: Session) -> int:
    """Переносит справочник из licenses.yml в таблицу `license` (для автодополнения)."""
    policy = get_license_policy()
    created = 0
    for entry in [*policy.allowed.values(), *policy.forbidden.values()]:
        row = session.execute(
            select(License).where(License.spdx_id == entry["spdx_id"])
        ).scalars().first()
        if row is None:
            row = License(spdx_id=entry["spdx_id"])
            session.add(row)
            created += 1
        row.name = entry.get("name")
        row.url = entry.get("url")
        row.notes = entry.get("notes")
        row.allowed = bool(entry.get("allowed"))
    session.flush()
    return created


def seed_demo_users(session: Session, *, password: str | None = None) -> list[User]:
    users: list[User] = []
    for username, full_name, roles, email in DEMO_USERS:
        user = session.execute(select(User).where(User.username == username)).scalars().first()
        if user is None:
            user = User(username=username, full_name=full_name, email=email, roles=roles)
            session.add(user)
        user.roles = roles
        user.full_name = full_name
        user.email = email
        if password:
            user.password_hash = hash_password(password)
            user.is_service = True
        users.append(user)
    session.flush()
    return users


def seed_demo_data(session: Session) -> dict[str, int]:
    """Демо-данные: одобренные пакеты и заявка, остановленная на карантине."""
    users = seed_demo_users(session)
    author = next(u for u in users if u.has_role("developer"))
    counts = {"approved": 0, "requests": 0}

    approved = [
        ("pypi", "requests", "2.31.0", "Apache-2.0"),
        ("pypi", "pydantic", "2.6.4", "MIT"),
        ("npm", "lodash", "4.17.21", "MIT"),
        ("go", "github.com/gin-gonic/gin", "v1.9.1", "MIT"),
        ("nuget", "Newtonsoft.Json", "13.0.3", "MIT"),
    ]
    for manager, name, version, spdx in approved:
        plugin = get_plugin(manager)
        ref = plugin.make_ref(name, version)
        pv, created = packages.get_or_create_version(session, ref)
        if created or pv.status != "approved":
            pv.status = "approved"
            pv.license_spdx = spdx
            pv.license_source = "registry"
            pv.published_at = utcnow() - timedelta(days=200)
            pv.approved_at = utcnow() - timedelta(days=30)
            pv.max_vuln_score = 0.0
            counts["approved"] += 1

    # Заявка в карантине: свежая версия, конвейер остановлен на шаге 2.
    plugin = get_plugin("pypi")
    ref = plugin.make_ref("httpx", "0.27.0")
    pv, _ = packages.get_or_create_version(session, ref)
    pv.status = "quarantined"
    pv.published_at = utcnow() - timedelta(days=3)
    pv.quarantine_until = utcnow() + timedelta(days=11)
    pv.license_spdx = "BSD-3-Clause"

    existing = session.execute(
        select(ModerationRequest).where(ModerationRequest.reason == "Демо-заявка: карантин")
    ).scalars().first()
    if existing is None:
        request = ModerationRequest(
            author_id=author.id,
            author_role="developer",
            manager="pypi",
            reason="Демо-заявка: карантин",
            status="quarantined",
            source="cli",
        )
        session.add(request)
        session.flush()
        item = RequestItem(
            request_id=request.id,
            package_version_id=pv.id,
            requested_name="httpx",
            requested_version="0.27.0",
            status="quarantined",
            current_step="quarantine",
            blocked_reason="Карантин 14 дн. не истёк",
            next_action="Проверка продолжится автоматически после окончания карантина.",
            waiting_since=utcnow(),
        )
        session.add(item)
        session.flush()
        ensure_steps(session, item)
        for step in item.steps:
            if step.step_code == "db_check":
                step.result, step.message = "pass", "В базе не найден, заявка принята к проверке."
            elif step.step_code == "blacklist":
                step.result, step.message = "pass", "Совпадений с правилами blacklist нет."
            elif step.step_code == "quarantine":
                step.result = "warn"
                step.message = "Версия опубликована 3 дн. назад, карантин 14 дн. не истёк."
            else:
                step.result = "skipped"
                step.message = "Не выполнялся: конвейер остановлен на шаге «Карантин»."
            step.finished_at = utcnow()
        recompute_request_status(session, request)
        counts["requests"] += 1

    audit.record(
        session,
        action="demo_data_seeded",
        entity_type="configuration",
        new_value=counts,
        source="cli",
    )
    session.flush()
    return counts


def import_package_list(
    session: Session,
    *,
    manager: str,
    lines: list[str],
    actor: User | None,
    origin: str,
) -> dict[str, int]:
    """Одноразовый импорт существующего `package_list.txt` как уже одобренных пакетов.

    Источник фиксируется в аудит-логе; после импорта файлы в GitLab не читаются.
    """
    plugin = get_plugin(manager)
    stats = {"imported": 0, "skipped": 0, "invalid": 0}
    for raw in lines:
        line = raw.split("#", 1)[0].strip()
        if not line:
            continue
        entry = plugin.try_parse_entry(line)
        if entry.ref is None:
            stats["invalid"] += 1
            continue
        version, created = packages.get_or_create_version(session, entry.ref)
        if not created and version.status == "approved":
            stats["skipped"] += 1
            continue
        version.status = "approved"
        version.approved_at = utcnow()
        version.license_source = version.license_source or "manual"
        version.status_reason = f"Импортирован из {origin} как ранее одобренный"
        stats["imported"] += 1
        audit.record(
            session,
            action="package_imported",
            entity_type="package_version",
            entity_id=version.id,
            actor=actor,
            new_value={
                "manager": manager,
                "name": entry.ref.display_name,
                "version": entry.ref.raw_version,
                "origin": origin,
            },
            source="cli",
            comment=f"Импорт из {origin}",
        )
    session.flush()
    return stats

File: backend-go/internal/seeds/seeds.go
// Package seeds — первичное заполнение: справочники, демо-стенд, разовый
// импорт существующих списков пакетов. Порт backend/app/services/seeds.py.
//
// Почему отдельный пакет, а не тело подкоманды: справочник менеджеров и
// лицензий заполняется не только из CLI. Его же перечитывает администратор
// через «Настройку» после правки licenses.yml, и разъехаться эти два пути не
// имеют права — иначе выпадающий список в карточке пакета зависел бы от того,
// каким способом справочник обновили в последний раз.
package seeds

import (
	"context"
	"fmt"
	"strings"
	"time"

	"moderation/internal/domain"
	"moderation/internal/policy"
	"moderation/internal/registry"
	"moderation/internal/repo"
)

// Service — заполнение справочников и демо-данных.
type Service struct {
	Repo     *repo.Repo
	Registry *registry.Registry
	// Policies — держатель политик, а не загруженный справочник: bootstrap
	// может идти следом за правкой licenses.yml, и брать надо текущее
	// содержимое файла, а не снимок, сделанный при сборке зависимостей.
	Policies *policy.Holder
	// Now — источник времени, подменяемый в тестах.
	Now func() time.Time
}

func (s *Service) now() time.Time {
	if s.Now != nil {
		return s.Now()
	}
	return time.Now().UTC()
}

// Managers переносит справочник менеджеров из плагинов в базу.
//
// Источник правды — плагины: заголовок и формат записи объявлены там, и
// вторая их копия в SQL-миграции разъехалась бы с разбором при первом же
// уточнении формулировки.
func (s *Service) Managers(ctx context.Context) (int, error) {
	plugins := s.Registry.Plugins()
	rows := make([]repo.ManagerSeed, 0, len(plugins))
	for _, p := range plugins {
		rows = append(rows, repo.ManagerSeed{
			Code: p.Code(), Title: p.Title(), EntryFormat: p.EntryFormat(),
		})
	}
	return s.Repo.SeedManagers(ctx, rows)
}

// ErrLicensesNotLoaded — справочник лицензий не прочитался.
//
// Отдельная ошибка, а не «заполнено 0 лицензий»: пустой справочник и
// непрочитанный файл выглядят одинаково по количеству строк, но означают
// противоположное. Молча записать в базу пустоту по нечитаемому файлу значило
// бы стереть уже заполненный справочник.
type ErrLicensesNotLoaded struct {
	Path   string
	Reason string
}

func (e *ErrLicensesNotLoaded) Error() string {
	return fmt.Sprintf("справочник лицензий %s не прочитан: %s", e.Path, e.Reason)
}

// Licenses переносит справочник SPDX из файла политики в таблицу.
//
// Зачем он в базе, если вердикт выносит файл: таблица нужна выпадающему списку
// в карточке пакета и внешнему ключу у заявления лицензии. Расхождение между
// ними не опасно, но список, в котором нет лицензии из файла, мешает
// разработчику её заявить.
func (s *Service) Licenses(ctx context.Context) (int, error) {
	p := s.Policies.Licenses()
	if p == nil {
		return 0, &ErrLicensesNotLoaded{Path: "(не задан)", Reason: "файл не настроен"}
	}
	if p.Failed() {
		return 0, &ErrLicensesNotLoaded{Path: p.Path, Reason: p.Err}
	}
	entries := append(p.SortedAllowed(), p.SortedForbidden()...)
	rows := make([]repo.LicenseSeed, 0, len(entries))
	for _, e := range entries {
		rows = append(rows, repo.LicenseSeed{
			SPDXID: e.SPDXID, Name: e.Name, URL: e.URL,
			Allowed: e.Allowed, Notes: e.Notes,
		})
	}
	return s.Repo.SeedLicenses(ctx, rows)
}

// ImportStats — исход импорта списка пакетов.
type ImportStats struct {
	// Imported — переведены в «одобрен».
	Imported int
	// Skipped — уже были одобрены, трогать нечего.
	Skipped int
	// Invalid — строка не разобралась форматом менеджера.
	Invalid int
	// InvalidLines — сами нераспознанные строки: без них человеку нечего
	// править. Ограничены, чтобы вывод не утонул в битом файле.
	InvalidLines []string
}

// maxInvalidShown — сколько нераспознанных строк показываем.
const maxInvalidShown = 20

// ImportPackageList разово вносит существующий package_list.txt как уже
// одобренные пакеты.
//
// Нужен один раз при заведении сервиса: до него список одобренного жил в
// файлах в GitLab, и без переноса каждый уже используемый пакет пришлось бы
// проводить через модерацию заново. После импорта файлы не читаются — источник
// фиксируется в журнале, чтобы через год было видно, откуда пакет взялся.
//
// Пакеты заводятся одобренными без прогона конвейера — это осознанно: решение
// по ним уже было принято людьми, и прогонять его повторно значило бы
// заблокировать половину работающего кода на время проверки.
func (s *Service) ImportPackageList(
	ctx context.Context, manager string, lines []string, actor *domain.User, origin string,
) (ImportStats, error) {
	var stats ImportStats
	plugin, err := s.Registry.Get(manager)
	if err != nil {
		return stats, err
	}
	now := s.now()
	reason := "Импортирован из " + origin + " как ранее одобренный"

	for _, raw := range lines {
		line := strings.TrimSpace(strings.SplitN(raw, "#", 2)[0])
		if line == "" {
			continue
		}
		ref, err := registry.ParseEntry(plugin, line)
		if err != nil {
			stats.Invalid++
			if len(stats.InvalidLines) < maxInvalidShown {
				stats.InvalidLines = append(stats.InvalidLines, line)
			}
			continue
		}
		pkg, err := s.Repo.GetOrCreatePackage(ctx, ref.Manager, ref.Name, ref.DisplayName)
		if err != nil {
			return stats, err
		}
		version, err := s.Repo.CreatePackageVersion(ctx, pkg.ID, ref.Version, ref.RawVersion)
		if err != nil {
			return stats, err
		}
		// Уже одобренную версию не трогаем: повторный запуск импорта по тому
		// же файлу обязан быть безвредным — иначе его боятся запускать, и
		// дописанные в список пакеты так и не попадают в базу.
		if version.Status == "approved" {
			stats.Skipped++
			continue
		}
		if err := s.Repo.MarkVersionImported(ctx, version.ID, reason, now); err != nil {
			return stats, err
		}
		stats.Imported++

		entityID := fmt.Sprintf("%d", version.ID)
		comment := "Импорт из " + origin
		entry := domain.AuditLog{
			Action: "package_imported", EntityType: "package_version", EntityID: &entityID,
			NewValue: map[string]any{
				"manager": ref.Manager, "name": ref.DisplayName,
				"version": ref.RawVersion, "origin": origin,
			},
			Source: "cli", Comment: &comment, CreatedAt: now,
		}
		if actor != nil {
			entry.ActorID = &actor.ID
			entry.ActorName = actor.Username
		} else {
			entry.ActorName = "CLI"
		}
		if err := s.Repo.InsertAuditLog(ctx, entry); err != nil {
			return stats, err
		}
	}
	return stats, nil
}
